libraries/botbuilder-dialogs-adaptive/src/actions/actionScope.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * @module botbuilder-dialogs-adaptive
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Dialog, DialogTurnResult, DialogContext, DialogReason, DialogConfiguration } from 'botbuilder-dialogs';
+
+/**
+ * @private
+ */
+const OFFSET_KEY: string = 'this.offset';
+
+export enum ActionScopeCommands {
+    goto = 'goto',
+    break = 'break',
+    continue = 'continue'
+}
+
+export interface ActionScopeResult {
+    actionScopeCommand: ActionScopeCommands;
+    actionId?: string;
+}
+
+export interface ActionScopeConfiguration extends DialogConfiguration {
+    /**
+     * Ordered set of actions to be executed within the scope.
+     */
+    actions?: Dialog[];
+} 
+
+export class ActionScope<O extends object = {}> extends Dialog<O> {
+
+    /**
+     * Creates a new `ActionBlock` instance.
+     */
+    constructor(actions?: Dialog[]) {
+        super();
+        if (actions) { this.actions = actions }
+    }
+
+    public actions: Dialog[] = [];
+
+    public getDependencies(): Dialog[] {
+        return this.actions;
+    }
+
+    protected onComputeId(): string {
+        const ids = this.actions.map(a => a.id);
+        return `ActionBlock[${this.hashedLabel(ids.join(','))}]`;
+    }
+
+    public configure(config: ActionScopeConfiguration): this {
+        return super.configure(config);
+    }
+
+    public async beginDialog(dc: DialogContext, options?: O): Promise<DialogTurnResult> {
+        if (this.actions.length > 0) {
+            return await this.beginAction(dc, 0);
+        } else {
+            return await dc.endDialog();
+        }
+    }
+
+    public async resumeDialog(dc: DialogContext, reason: DialogReason, result?: any): Promise<DialogTurnResult> {
+        // Check for action block command
+        const nextOffset = dc.state.getValue(OFFSET_KEY) + 1;
+        if (typeof result === 'object' && result.hasOwnProperty('actionScopeCommand')) {
+            switch ((result as ActionScopeResult).actionScopeCommand) {
+                case ActionScopeCommands.goto:
+                    return await this.gotoAction(dc, (result as ActionScopeResult).actionId);
+                case ActionScopeCommands.continue:
+                    return await this.onContinue(dc);
+                case ActionScopeCommands.break:
+                default:
+                    return await this.onBreak(dc);
+            }
```

Comment:
I pulled this into virtual OnActionScopeResult(ActionScopeResult ) method so that we can extend the processing of the commands by overriding the processing of them